charmcraft/charm_builder.py

```diff
@@ -224,21 +229,23 @@ def handle_dependencies(self):
 
         # virtualenv with other dependencies (if any)
         if self.requirement_paths:
-            _process_run(["pip3", "--version"])
-
-            venvpath = self.buildpath / VENV_DIRNAME
-            cmd = [
-                "pip3",
-                "install",  # base command
-                "--target={}".format(venvpath),  # put all the resulting files in that specific dir
-            ]
+            staging_venv_dir = self.charmdir / STAGING_VENV_DIRNAME
+            _process_run(["python3", "-m", "venv", str(staging_venv_dir)])
```

Comment:
It was originally implemented using `venv.create(symlinks=True, with_pip=True)`, however for some reason ensurepip fails silently when invoked from the current interpreter and the resulting virtualenv has no `bin/pip` or `bin/pip3`. Calling as an external process fixes this issue.

Filenames with spaces should be handled by Python's `Popen` because the argument is passed as a single `argv` member instead of being split by the shell into multiple arguments (unlike `popen(3)` that takes a command string and invokes a shell). On Windows `Popen` calls `CreateProcess()` which takes a single command string, but [according to the documentation](https://docs.python.org/3/library/subprocess.html#subprocess.Popen) it should be converted taking whitespaces and other special characters into consideration (but please verify the documentation to confirm that my interpretation is correct):

> On Windows, if args is a sequence, it will be converted to a string in a manner described in [Converting an argument sequence to a string on Windows](https://docs.python.org/3/library/subprocess.html#converting-argument-sequence). This is because the underlying CreateProcess() operates on strings.

This simple test works as expected on POSIX:
```
>>> import subprocess
>>> x=subprocess.Popen(["ls", "/home/claudio/test with spaces"])
>>> foo
```

Reply:
Ok!